scripts/dss_deployer.py

```diff
@@ -0,0 +1,403 @@
+parser.add_argument('action', type=str, choices=['deploy', 'destroy'])
+parser.add_argument('--no-test-buckets', dest='create_test_buckets', action='store_false', default=True)
+parser.add_argument('--accept-defaults', dest='accept_defaults', action='store_true', default=False)
+parser.add_argument('--virtualenv', dest='virtualenv', default=os.path.join(pkg_root, 'virtualenv'))
+
+
+def no_action(msg):
+    print(f'NO ACTION {msg}')
+
+
+def find(cmd):
+    path = shutil.which(cmd)
+
+    if not os.path.isfile(path):
+        raise Exception(f'Command not fond: {cmd}')
+
+    return path
+
+
+def request_user_input(key, default=None):
+    if default is None:
+        return input(f'{key.upper()}=')
+    else:
+        if args.accept_defaults:
+            print(f'setting {key.upper()}={default}')
+            return default
+        return input(f'{key.upper()}[default={default}]=') or default
+
+
+def request_user_continuation(msg, default=True):
+    print(msg)
+
+    if args.accept_defaults:
+        return
+
+    if default:
+        resp = input('Proceed (Y/n)').lower() or 'y'
+    else:
+        resp = input('Proceed (y/N)').lower() or 'n'
+
+    if 'y' != resp:
+        sys.exit(1)
+
+
+class Command:
+    count = 0
+
+    def __init__(self, cmd=None):
+        self.cmd = cmd
+
+    @classmethod
+    def call(cls, *args, quiet=False):
+        if not quiet:
+            cmd_string = ' '.join(args)
+            Command.count += 1
+            print(f'{Command.count} {cmd_string}')
+        out = subprocess.run(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, encoding='utf-8')
+        try:
+            out.check_returncode()
+        except subprocess.CalledProcessError:
+            if not quiet:
+                print(f'\t{out.stderr}')
+        return out.stdout.strip()
+
+    def __call__(self, *args, **kwargs):
+        return Command.call(*[self.cmd, *args], **kwargs)
+
+    def quiet(self, *args):
+        return self(*args, quiet=True)
+
+    @classmethod
+    def check_python_version(cls):
+        python = find('python')
+        out = Command.call('python', '-c', 'import sys;print(sys.version)', quiet=True)
+        if not out.startswith('3.6'):
+            raise Exception('Python version must be >= 3.6')
+
+    @classmethod
+    def setup_virtualenv(self, path):
+        if not os.path.exists(path):
+            Command.call(find('virtualenv'), '-p', python, self.virtualenv_path)
+
+    @classmethod
+    def issue_shell_commands(cls, *cmds):
+        cmd = ';'.join(cmds)
+        proc = subprocess.Popen([cmd],
+            stdout=sys.stdout.buffer,
+            stderr=sys.stderr.buffer,
+            stdin=sys.stdin.buffer,
+            shell=True,
+            executable='/bin/bash',
+            encoding='utf-8',
+            errors='backslashreplace',
+            bufsize=1,
+            universal_newlines=True,
+        )
+
+        proc.communicate('')
+
+
+aws = Command(find('aws'))
+gcloud = Command(find('gcloud'))
```

Comment:
I'm not 100% sure about gcloud, but for aws I would consider using boto3 directly instead. It makes things less painful down the road when you inevitably accrue more logic for handling errors and other conditions.

Reply:
@kislyuk My original intent was the avoidance of tools unavailable on the command line. Todays discussion with @hannes-ucsc will cause some refactoring, however, and I may head down the boto path.